File: app/src/main/java/com/mohsenoid/rickandmorty/data/db/DbImpl.java
package com.mohsenoid.rickandmorty.data.db;

import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;

import androidx.annotation.VisibleForTesting;

import com.mohsenoid.rickandmorty.data.Serializer;
import com.mohsenoid.rickandmorty.data.db.dto.DbCharacterModel;
import com.mohsenoid.rickandmorty.data.db.dto.DbEpisodeModel;
import com.mohsenoid.rickandmorty.data.db.dto.DbLocationModel;
import com.mohsenoid.rickandmorty.data.db.dto.DbOriginModel;
import com.mohsenoid.rickandmorty.data.exception.NoOfflineDataException;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;

public class DbImpl extends SQLiteOpenHelper implements Db {

    @VisibleForTesting
    public static DbImpl instance;

    private final SQLiteDatabase db;

    private DbImpl(Context context) {
        super(context, DbConstants.DATABASE_NAME, null, DbConstants.DATABASE_VERSION);
        db = getWritableDatabase();
    }

    public static synchronized DbImpl getInstance(Context context) {
        if (instance == null)
            instance = new DbImpl(context);

        return instance;
    }

    @Override
    synchronized public void onCreate(SQLiteDatabase db) {
        createEpisodeTable(db);
        createCharacterTable(db);
    }

    synchronized private void createEpisodeTable(SQLiteDatabase db) {
        String sql = String
                .format("CREATE TABLE IF NOT EXISTS %s (%s INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL," +
                                "%s TEXT, %s TEXT, %s TEXT, %s TEXT, %s TEXT, %s TEXT)",
                        DbConstants.Episode.TABLE_NAME,
                        DbConstants.Episode.ID,
                        DbConstants.Episode.NAME,
                        DbConstants.Episode.AIR_DATE,
                        DbConstants.Episode.EPISODE,
                        DbConstants.Episode.CHARACTER_IDS,
                        DbConstants.Episode.URL,
                        DbConstants.Episode.CREATED);
        db.execSQL(sql);
    }

    synchronized private void createCharacterTable(SQLiteDatabase db) {
        String sql = String
                .format("CREATE TABLE IF NOT EXISTS %s (%s INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL," +
                                "%s TEXT, %s TEXT, %s INTEGER, %s TEXT, %s TEXT, %s TEXT, %s TEXT, %s TEXT, %s TEXT, %s TEXT, %s TEXT, %s TEXT, %s TEXT, %s TEXT, %s INTEGER DEFAULT 0)",
                        DbConstants.Character.TABLE_NAME,
                        DbConstants.Character.ID,
                        DbConstants.Character.NAME,
                        DbConstants.Character.STATUS,
                        DbConstants.Character.IS_STATUS_ALIVE,
                        DbConstants.Character.SPECIES,
                        DbConstants.Character.TYPE,
                        DbConstants.Character.GENDER,
                        DbConstants.Character.ORIGIN_NAME,
                        DbConstants.Character.ORIGIN_URL,
                        DbConstants.Character.LOCATION_NAME,
                        DbConstants.Character.LOCATION_URL,
                        DbConstants.Character.IMAGE,
                        DbConstants.Character.EPISODE,
                        DbConstants.Character.URL,
                        DbConstants.Character.CREATED,
                        DbConstants.Character.KILLED_BY_USER);
        db.execSQL(sql);
    }

    @Override
    synchronized public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
        dropEpisodeTable(db);
        dropCharacterTable(db);

        onCreate(db);
    }

    synchronized private void dropEpisodeTable(SQLiteDatabase db) {
        String sql = String.format("DROP TABLE IF EXISTS %s", DbConstants.Episode.TABLE_NAME);
        db.execSQL(sql);
    }

    synchronized private void dropCharacterTable(SQLiteDatabase db) {
        String sql = String.format("DROP TABLE IF EXISTS %s", DbConstants.Character.TABLE_NAME);
        db.execSQL(sql);
    }

    @Override
    synchronized public void insertEpisode(DbEpisodeModel episode) {
        ContentValues contentValues = new ContentValues();
        contentValues.put(DbConstants.Episode.ID, episode.getId());
        contentValues.put(DbConstants.Episode.NAME, episode.getName());
        contentValues.put(DbConstants.Episode.AIR_DATE, episode.getAirDate());
        contentValues.put(DbConstants.Episode.EPISODE, episode.getEpisode());
        contentValues.put(DbConstants.Episode.CHARACTER_IDS, episode.getSerializedCharacterIds());
        contentValues.put(DbConstants.Episode.URL, episode.getUrl());
        contentValues.put(DbConstants.Episode.CREATED, episode.getCreated());

        Cursor selectEpisodeCursor = getSelectEpisodeCursor(episode.getId());
        if (selectEpisodeCursor.moveToFirst()) {
            String whereClause = DbConstants.Episode.ID + "=?";
            String[] args = {selectEpisodeCursor.getInt(selectEpisodeCursor.getColumnIndexOrThrow(DbConstants.Episode.ID)) + ""};
            db.update(DbConstants.Episode.TABLE_NAME, contentValues, whereClause, args);
        } else {
            db.insert(DbConstants.Episode.TABLE_NAME, DbConstants.Episode.ID, contentValues);
        }
    }

    synchronized private Cursor getSelectEpisodeCursor(int episodeId) {
        String sql = String.format(
                "SELECT * FROM %s WHERE %s = %s",
                DbConstants.Episode.TABLE_NAME,
                DbConstants.Episode.ID,
                episodeId);
        return db.rawQuery(sql, null);
    }

    @Override
    synchronized public List<DbEpisodeModel> queryAllEpisodes(int page) {
        List<DbEpisodeModel> episodes = new ArrayList<>();

        String sql = String.format("SELECT * FROM %s LIMIT %s OFFSET %s", DbConstants.Episode.TABLE_NAME, DbConstants.PAGE_SIZE, calculatePageOffset(page));
        try (Cursor cursor = db.rawQuery(sql, null)) {
            while (cursor.moveToNext()) {
                int id = cursor.getInt(cursor.getColumnIndexOrThrow(DbConstants.Episode.ID));
                String name = cursor.getString(cursor.getColumnIndexOrThrow(DbConstants.Episode.NAME));
                String airDate = cursor.getString(cursor.getColumnIndexOrThrow(DbConstants.Episode.AIR_DATE));
                String episode = cursor.getString(cursor.getColumnIndexOrThrow(DbConstants.Episode.EPISODE));
                String characterIds = cursor.getString(cursor.getColumnIndexOrThrow(DbConstants.Episode.CHARACTER_IDS));
                String url = cursor.getString(cursor.getColumnIndexOrThrow(DbConstants.Episode.URL));
                String created = cursor.getString(cursor.getColumnIndexOrThrow(DbConstants.Episode.CREATED));

                DbEpisodeModel newEpisode = new DbEpisodeModel(id, name, airDate, episode, characterIds, url, created);
                episodes.add(newEpisode);
            }
        }

        return episodes;
    }

    @Override
    synchronized public void insertCharacter(DbCharacterModel character) {
        ContentValues contentValues = new ContentValues();
        contentValues.put(DbConstants.Character.ID, character.getId());
        contentValues.put(DbConstants.Character.NAME, character.getName());
        contentValues.put(DbConstants.Character.STATUS, character.getStatus());
        contentValues.put(DbConstants.Character.IS_STATUS_ALIVE, character.getStatusAlive());
        contentValues.put(DbConstants.Character.SPECIES, character.getSpecies());
        contentValues.put(DbConstants.Character.TYPE, character.getType());
        contentValues.put(DbConstants.Character.GENDER, character.getGender());
        contentValues.put(DbConstants.Character.ORIGIN_NAME, character.getOrigin().getName());
        contentValues.put(DbConstants.Character.ORIGIN_URL, character.getOrigin().getUrl());
        contentValues.put(DbConstants.Character.LOCATION_NAME, character.getLocation().getName());
        contentValues.put(DbConstants.Character.LOCATION_URL, character.getLocation().getUrl());
        contentValues.put(DbConstants.Character.IMAGE, character.getImage());
        contentValues.put(DbConstants.Character.EPISODE, character.getSerializedEpisodes());
        contentValues.put(DbConstants.Character.URL, character.getUrl());
        contentValues.put(DbConstants.Character.CREATED, character.getCreated());

        Cursor selectEpisodeCursor = getSelectCharacterCursor(character.getId());
        if (selectEpisodeCursor.moveToFirst()) {
            String whereClause = DbConstants.Character.ID + "=?";
            String[] args = {selectEpisodeCursor.getInt(selectEpisodeCursor.getColumnIndexOrThrow(DbConstants.Episode.ID)) + ""};
            db.update(DbConstants.Character.TABLE_NAME, contentValues, whereClause, args);
        } else {
            db.insert(DbConstants.Character.TABLE_NAME, DbConstants.Character.ID, contentValues);
        }
    }

    synchronized private Cursor getSelectCharacterCursor(int characterId) {
        String sql = String.format(
                "SELECT * FROM %s WHERE %s = %s",
                DbConstants.Character.TABLE_NAME,
                DbConstants.Character.ID,
                characterId);
        return db.rawQuery(sql, null);
    }

    @Override
    synchronized public List<DbCharacterModel> queryCharactersByIds(List<Integer> characterIds) {
        if (characterIds == null) return null;

        List<DbCharacterModel> characters = new ArrayList<>();
        if (characterIds.size() == 0) return characters;

        String sql = String.format("SELECT * FROM %s WHERE %s IN (%s)", DbConstants.Character.TABLE_NAME, DbConstants.Character.ID, Serializer.serializeIntegerList(characterIds));
        try (Cursor cursor = db.rawQuery(sql, null)) {
            while (cursor.moveToNext()) {
                DbCharacterModel newCharacter = getCursorCharacterDetails(cursor);
                characters.add(newCharacter);
            }
        }

        return characters;
    }

    private int calculatePageOffset(int page) {
        return (page - 1) * DbConstants.PAGE_SIZE;
    }

    @Override
    synchronized public DbCharacterModel queryCharacter(int characterId) {
        DbCharacterModel character = null;

        try (Cursor cursor = getSelectCharacterCursor(characterId)) {
            if (cursor.moveToFirst()) {
                character = getCursorCharacterDetails(cursor);
            }
        }

        return character;
    }

    @NotNull
    private DbCharacterModel getCursorCharacterDetails(Cursor cursor) {
        int id = cursor.getInt(cursor.getColumnIndexOrThrow(DbConstants.Character.ID));
        String name = cursor.getString(cursor.getColumnIndexOrThrow(DbConstants.Character.NAME));
        String status = cursor.getString(cursor.getColumnIndexOrThrow(DbConstants.Character.STATUS));
        boolean isStatusAlive = cursor.getInt(cursor.getColumnIndexOrThrow(DbConstants.Character.IS_STATUS_ALIVE)) == 1;
        String species = cursor.getString(cursor.getColumnIndexOrThrow(DbConstants.Character.SPECIES));
        String type = cursor.getString(cursor.getColumnIndexOrThrow(DbConstants.Character.TYPE));
        String gender = cursor.getString(cursor.getColumnIndexOrThrow(DbConstants.Character.GENDER));
        String originName = cursor.getString(cursor.getColumnIndexOrThrow(DbConstants.Character.ORIGIN_NAME));
        String originUrl = cursor.getString(cursor.getColumnIndexOrThrow(DbConstants.Character.ORIGIN_URL));
        String locationName = cursor.getString(cursor.getColumnIndexOrThrow(DbConstants.Character.LOCATION_NAME));
        String locationUrl = cursor.getString(cursor.getColumnIndexOrThrow(DbConstants.Character.LOCATION_URL));
        String image = cursor.getString(cursor.getColumnIndexOrThrow(DbConstants.Character.IMAGE));
        String episode = cursor.getString(cursor.getColumnIndexOrThrow(DbConstants.Character.EPISODE));
        String url = cursor.getString(cursor.getColumnIndexOrThrow(DbConstants.Character.URL));
        String created = cursor.getString(cursor.getColumnIndexOrThrow(DbConstants.Character.CREATED));
        boolean isKilledByUser = cursor.getInt(cursor.getColumnIndexOrThrow(DbConstants.Character.KILLED_BY_USER)) == 1;

        DbOriginModel characterOrigin = new DbOriginModel(originName, originUrl);
        DbLocationModel characterLocation = new DbLocationModel(locationName, locationUrl);

        return new DbCharacterModel(id, name, status, isStatusAlive, species, type, gender, characterOrigin, characterLocation, image, episode, url, created, isKilledByUser);
    }

    @Override
    public void killCharacter(int characterId) throws NoOfflineDataException {
        Cursor selectEpisodeCursor = getSelectCharacterCursor(characterId);
        if (selectEpisodeCursor.moveToFirst()) {
            ContentValues contentValues = new ContentValues();
            contentValues.put(DbConstants.Character.KILLED_BY_USER, 1);

            String whereClause = DbConstants.Character.ID + "=?";
            String[] args = {selectEpisodeCursor.getInt(selectEpisodeCursor.getColumnIndexOrThrow(DbConstants.Episode.ID)) + ""};
            db.update(DbConstants.Character.TABLE_NAME, contentValues, whereClause, args);
        } else {
            throw new NoOfflineDataException();
        }
    }
}

File: app/src/main/kotlin/com/mohsenoid/rickandmorty/data/db/DbImpl.kt
package com.mohsenoid.rickandmorty.data.db

import android.content.ContentValues
import android.content.Context
import android.database.Cursor
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteOpenHelper
import com.mohsenoid.rickandmorty.data.Serializer
import com.mohsenoid.rickandmorty.data.db.dto.DbCharacterModel
import com.mohsenoid.rickandmorty.data.db.dto.DbEpisodeModel
import com.mohsenoid.rickandmorty.data.db.dto.DbLocationModel
import com.mohsenoid.rickandmorty.data.db.dto.DbOriginModel
import com.mohsenoid.rickandmorty.data.exception.NoOfflineDataException
import java.util.ArrayList

class DbImpl(context: Context) :
    SQLiteOpenHelper(context, DbConstants.DATABASE_NAME, null, DbConstants.DATABASE_VERSION), Db {

    private val db: SQLiteDatabase = writableDatabase

    @Synchronized
    override fun onCreate(db: SQLiteDatabase) {
        createEpisodeTable(db)
        createCharacterTable(db)
    }

    @Synchronized
    private fun createEpisodeTable(db: SQLiteDatabase) {
        val sql = "CREATE TABLE IF NOT EXISTS ${DbConstants.Episode.TABLE_NAME} " +
            "(" +
            "${DbConstants.Episode.ID} INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL," +
            "${DbConstants.Episode.NAME} TEXT, " +
            "${DbConstants.Episode.AIR_DATE} TEXT, " +
            "${DbConstants.Episode.EPISODE} TEXT, " +
            "${DbConstants.Episode.CHARACTER_IDS} TEXT, " +
            "${DbConstants.Episode.URL} TEXT, " +
            "${DbConstants.Episode.CREATED} TEXT" +
            ")"
        db.execSQL(sql)
    }

    @Synchronized
    private fun createCharacterTable(db: SQLiteDatabase) {
        val sql = "CREATE TABLE IF NOT EXISTS ${DbConstants.Character.TABLE_NAME} " +
            "(" +
            "${DbConstants.Character.ID} INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL, " +
            "${DbConstants.Character.NAME} TEXT, " +
            "${DbConstants.Character.STATUS} TEXT, " +
            "${DbConstants.Character.IS_STATUS_ALIVE} INTEGER, " +
            "${DbConstants.Character.SPECIES} TEXT, " +
            "${DbConstants.Character.TYPE} TEXT, " +
            "${DbConstants.Character.GENDER} TEXT, " +
            "${DbConstants.Character.ORIGIN_NAME} TEXT, " +
            "${DbConstants.Character.ORIGIN_URL} TEXT, " +
            "${DbConstants.Character.LOCATION_NAME} TEXT, " +
            "${DbConstants.Character.LOCATION_URL} TEXT, " +
            "${DbConstants.Character.IMAGE} TEXT, " +
            "${DbConstants.Character.EPISODE} TEXT, " +
            "${DbConstants.Character.URL} TEXT, " +
            "${DbConstants.Character.CREATED} TEXT, " +
            "${DbConstants.Character.KILLED_BY_USER} INTEGER DEFAULT 0" +
            ")"
        db.execSQL(sql)
    }

    @Synchronized
    override fun onUpgrade(db: SQLiteDatabase, oldVersion: Int, newVersion: Int) {
        dropTable(db, DbConstants.Episode.TABLE_NAME)
        dropTable(db, DbConstants.Character.TABLE_NAME)
        onCreate(db)
    }

    @Synchronized
    private fun dropTable(db: SQLiteDatabase, tableName: String) {
        val sql = "DROP TABLE IF EXISTS $tableName"
        db.execSQL(sql)
    }

    @Synchronized
    override fun insertEpisode(episode: DbEpisodeModel) {
        val contentValues = ContentValues()

        contentValues.put(DbConstants.Episode.ID, episode.id)
        contentValues.put(DbConstants.Episode.NAME, episode.name)
        contentValues.put(DbConstants.Episode.AIR_DATE, episode.airDate)
        contentValues.put(DbConstants.Episode.EPISODE, episode.episode)
        contentValues.put(DbConstants.Episode.CHARACTER_IDS, episode.serializedCharacterIds)
        contentValues.put(DbConstants.Episode.URL, episode.url)
        contentValues.put(DbConstants.Episode.CREATED, episode.created)

        val selectEpisodeCursor = getSelectEpisodeCursor(episode.id)
        if (selectEpisodeCursor.moveToFirst()) {
            val whereClause = "${DbConstants.Episode.ID}=?"
            val args = arrayOf(
                selectEpisodeCursor.getInt(
                    selectEpisodeCursor.getColumnIndexOrThrow(DbConstants.Episode.ID)
                ).toString() + ""
            )
            db.update(DbConstants.Episode.TABLE_NAME, contentValues, whereClause, args)
        } else {
            db.insert(DbConstants.Episode.TABLE_NAME, DbConstants.Episode.ID, contentValues)
        }
    }

    @Synchronized
    private fun getSelectEpisodeCursor(episodeId: Int): Cursor {
        val sql = "SELECT * FROM ${DbConstants.Episode.TABLE_NAME} " +
            "WHERE ${DbConstants.Episode.ID} = $episodeId"
        return db.rawQuery(sql, null)
    }

    @Synchronized
    override fun queryAllEpisodes(page: Int): List<DbEpisodeModel> {
        val episodes = ArrayList<DbEpisodeModel>()

        val sql = "SELECT * FROM ${DbConstants.Episode.TABLE_NAME} " +
            "LIMIT ${DbConstants.PAGE_SIZE} " +
            "OFFSET ${calculatePageOffset(page)}"

        db.rawQuery(sql, null).use { cursor ->
            while (cursor.moveToNext()) {
                val id = cursor.getInt(cursor.getColumnIndexOrThrow(DbConstants.Episode.ID))
                val name =
                    cursor.getString(cursor.getColumnIndexOrThrow(DbConstants.Episode.NAME))
                val airDate =
                    cursor.getString(cursor.getColumnIndexOrThrow(DbConstants.Episode.AIR_DATE))
                val episode =
                    cursor.getString(cursor.getColumnIndexOrThrow(DbConstants.Episode.EPISODE))
                val characterIds =
                    cursor.getString(cursor.getColumnIndexOrThrow(DbConstants.Episode.CHARACTER_IDS))
                val url =
                    cursor.getString(cursor.getColumnIndexOrThrow(DbConstants.Episode.URL))
                val created =
                    cursor.getString(cursor.getColumnIndexOrThrow(DbConstants.Episode.CREATED))

                val newEpisode = DbEpisodeModel(
                    id = id,
                    name = name,
                    airDate = airDate,
                    episode = episode,
                    serializedCharacterIds = characterIds,
                    url = url,
                    created = created
                )

                episodes.add(newEpisode)
            }
        }

        return episodes
    }

    @Synchronized
    override fun insertCharacter(character: DbCharacterModel) {
        val contentValues = ContentValues()

        contentValues.put(DbConstants.Character.ID, character.id)
        contentValues.put(DbConstants.Character.NAME, character.name)
        contentValues.put(DbConstants.Character.STATUS, character.status)
        contentValues.put(DbConstants.Character.IS_STATUS_ALIVE, character.statusAlive)
        contentValues.put(DbConstants.Character.SPECIES, character.species)
        contentValues.put(DbConstants.Character.TYPE, character.type)
        contentValues.put(DbConstants.Character.GENDER, character.gender)
        contentValues.put(DbConstants.Character.ORIGIN_NAME, character.origin.name)
        contentValues.put(DbConstants.Character.ORIGIN_URL, character.origin.url)
        contentValues.put(DbConstants.Character.LOCATION_NAME, character.location.name)
        contentValues.put(DbConstants.Character.LOCATION_URL, character.location.url)
        contentValues.put(DbConstants.Character.IMAGE, character.image)
        contentValues.put(DbConstants.Character.EPISODE, character.serializedEpisodes)
        contentValues.put(DbConstants.Character.URL, character.url)
        contentValues.put(DbConstants.Character.CREATED, character.created)

        val selectEpisodeCursor = getSelectCharacterCursor(character.id)
        if (selectEpisodeCursor.moveToFirst()) {
            val whereClause = "${DbConstants.Character.ID}=?"
            val args = arrayOf(
                selectEpisodeCursor.getInt(
                    selectEpisodeCursor.getColumnIndexOrThrow(DbConstants.Episode.ID)
                ).toString() + ""
            )
            db.update(DbConstants.Character.TABLE_NAME, contentValues, whereClause, args)
        } else {
            db.insert(DbConstants.Character.TABLE_NAME, DbConstants.Character.ID, contentValues)
        }
    }

    @Synchronized
    private fun getSelectCharacterCursor(characterId: Int): Cursor {
        val sql = "SELECT * FROM ${DbConstants.Character.TABLE_NAME} " +
            "WHERE ${DbConstants.Character.ID} = $characterId"
        return db.rawQuery(sql, null)
    }

    @Synchronized
    override fun queryCharactersByIds(characterIds: List<Int>): List<DbCharacterModel> {
        val characters = ArrayList<DbCharacterModel>()

        if (characterIds.isEmpty()) return characters

        val serializedCharacterIds = Serializer.serializeIntegerList(characterIds)
        val sql = "SELECT * FROM ${DbConstants.Character.TABLE_NAME} " +
            "WHERE ${DbConstants.Character.ID} IN ($serializedCharacterIds)"
        db.rawQuery(sql, null).use { cursor ->
            while (cursor.moveToNext()) {
                val newCharacter = getCursorCharacterDetails(cursor)
                characters.add(newCharacter)
            }
        }
        return characters
    }

    private fun calculatePageOffset(page: Int): Int {
        return (page - 1) * DbConstants.PAGE_SIZE
    }

    @Synchronized
    override fun queryCharacter(characterId: Int): DbCharacterModel? {
        var character: DbCharacterModel? = null

        getSelectCharacterCursor(characterId).use { cursor ->
            if (cursor.moveToFirst()) {
                character = getCursorCharacterDetails(cursor)
            }
        }
        return character
    }

    private fun getCursorCharacterDetails(cursor: Cursor): DbCharacterModel {
        val id =
            cursor.getInt(cursor.getColumnIndexOrThrow(DbConstants.Character.ID))
        val name =
            cursor.getString(cursor.getColumnIndexOrThrow(DbConstants.Character.NAME))
        val status =
            cursor.getString(cursor.getColumnIndexOrThrow(DbConstants.Character.STATUS))
        val isStatusAlive =
            cursor.getInt(cursor.getColumnIndexOrThrow(DbConstants.Character.IS_STATUS_ALIVE)) == 1
        val species =
            cursor.getString(cursor.getColumnIndexOrThrow(DbConstants.Character.SPECIES))
        val type =
            cursor.getString(cursor.getColumnIndexOrThrow(DbConstants.Character.TYPE))
        val gender =
            cursor.getString(cursor.getColumnIndexOrThrow(DbConstants.Character.GENDER))
        val originName =
            cursor.getString(cursor.getColumnIndexOrThrow(DbConstants.Character.ORIGIN_NAME))
        val originUrl =
            cursor.getString(cursor.getColumnIndexOrThrow(DbConstants.Character.ORIGIN_URL))
        val locationName =
            cursor.getString(cursor.getColumnIndexOrThrow(DbConstants.Character.LOCATION_NAME))
        val locationUrl =
            cursor.getString(cursor.getColumnIndexOrThrow(DbConstants.Character.LOCATION_URL))
        val image =
            cursor.getString(cursor.getColumnIndexOrThrow(DbConstants.Character.IMAGE))
        val episode =
            cursor.getString(cursor.getColumnIndexOrThrow(DbConstants.Character.EPISODE))
        val url =
            cursor.getString(cursor.getColumnIndexOrThrow(DbConstants.Character.URL))
        val created =
            cursor.getString(cursor.getColumnIndexOrThrow(DbConstants.Character.CREATED))
        val isKilledByUser =
            cursor.getInt(cursor.getColumnIndexOrThrow(DbConstants.Character.KILLED_BY_USER)) == 1
        val characterOrigin =
            DbOriginModel(originName, originUrl)
        val characterLocation =
            DbLocationModel(locationName, locationUrl)

        return DbCharacterModel(
            id = id,
            name = name,
            status = status,
            statusAlive = isStatusAlive,
            species = species,
            type = type,
            gender = gender,
            origin = characterOrigin,
            location = characterLocation,
            image = image,
            serializedEpisodes = episode,
            url = url,
            created = created,
            killedByUser = isKilledByUser
        )
    }

    override fun killCharacter(characterId: Int) {
        val selectEpisodeCursor = getSelectCharacterCursor(characterId)

        if (selectEpisodeCursor.moveToFirst()) {
            val contentValues = ContentValues()
            contentValues.put(DbConstants.Character.KILLED_BY_USER, 1)
            val whereClause = "${DbConstants.Character.ID}=?"
            val args = arrayOf(
                selectEpisodeCursor.getInt(
                    selectEpisodeCursor.getColumnIndexOrThrow(DbConstants.Episode.ID)
                ).toString()
            )

            db.update(DbConstants.Character.TABLE_NAME, contentValues, whereClause, args)
        } else {
            throw NoOfflineDataException()
        }
    }
}
